src/common/Utilities/Util.cpp

```diff
@@ -421,16 +421,21 @@ std::wstring GetMainPartOfName(std::wstring const& wname, uint32 declension)
 
     std::size_t const thisLen = wname.length();
     std::array<std::wstring const*, 7> const& endings = dropEnds[declension];
-    for (auto itr = endings.begin(), end = endings.end(); (itr != end) && *itr; ++itr)
+    for (const std::wstring* endingPtr : endings)
     {
-        std::wstring const& ending = **itr;
+        if (endingPtr == nullptr)
+        {
+            break;
+        }
+
+        std::wstring const& ending = *endingPtr;
         std::size_t const endLen = ending.length();
         if (endLen > thisLen)
         {
             continue;
         }
 
-        if (wname.substr(thisLen - endLen, thisLen) == ending)
```

Comment:
The comparison logic might be flawed because the substring method is used incorrectly. It should compare the correct substring of the name with the ending.